### PR TITLE
storage: Add Cloudflare as a CDN provider for an S3 backed storage (PROJQUAY-3699)

### PR DESCRIPTION
**Pull request by syed**
This adds CloudFlare as a CDN provider for quay for any storage backed
by S3. This requires a worker script that needs to be setup seperately
on CloudFlare. More details on the worker at
https://github.com/quay/quay-cloudflare-cdn-worker

### CODE REVIEW

File: storage/cloudflarestorage.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        super(CloudFlareS3Storage, self).__init__(context, storage_path, s3_bucket, *args, **kwargs)
+
+        self.cloudflare_domain = cloudflare_domain
+        self.cloudflare_privatekey = self._load_private_key(cloudflare_privatekey_filename)
+
+    def get_direct_download_url(
+        self, path, request_ip=None, expires_in=60, requires_cors=False, head=False
+    ):
+        # If CloudFront could not be loaded, fall back to normal S3.
+        s3_presigned_url = super(CloudFlareS3Storage, self).get_direct_download_url(
```
Comment (kleesc) [inline storage/cloudflarestorage.py:40]: Put this inside the `if` block, that way it doesn't have to sign _every_ request.

**Review by kleesc — CHANGES_REQUESTED** [review]

File: storage/cloudflarestorage.py (hunk 1 above)
Comment (kleesc) [reply storage/cloudflarestorage.py:40]: Just saw that this is still used even if you're able to sign with CloudFlare keys.
Then the comment `If CloudFront could not be loaded, fall back to normal S3.` doesn't really hold right?

Comment (syed) [reply storage/cloudflarestorage.py:40]: This is basically a graceful fallback. We would always set `cloudflare_privatekey`. I could instead throw an error here. I was just following how CloudFront has implemented it

Comment (kleesc) [reply storage/cloudflarestorage.py:40]: Ok. I guess I was confused since you were building the cloudflare_url based off the s3 one...

**Review by kleesc — APPROVED** [review]

**Pull request merged**